soc/arm/ti_simplelink/cc13x2_cc26x2/power.c

```diff
@@ -100,8 +100,18 @@ __weak void pm_power_state_set(struct pm_state_info info)
 		break;
 
 	case PM_STATE_STANDBY:
-		/* go to standby mode */
-		Power_sleep(PowerCC26XX_STANDBY);
+		/* NOTE: HAL constraints are checked here to cover the case
+		 * where Power_init() sets the standby constraint while clock
+		 * gets ready under certain configurations. This should have
+		 * been initially solved by implementing a proper clock
+		 * controller driver, so that the Zephyr PM APIs are used. HAL
+		 * code that has such own internal behavior (including managing
+		 * interrupts) should not be allowed in the future.
+		 */
+		constraints = Power_getConstraintMask();
+		if ((constraints & BIT(PowerCC26XX_DISALLOW_STANDBY)) == 0U) {
+			Power_sleep(PowerCC26XX_STANDBY);
+		}
```

Comment:
I agree with most arguments, but I believe checking the HAL constraint
here is not correct. The policy has already returned this state and
the subsystem will have suspended all devices necessarily :/ The
place to check for constraints is the policy.

I think if we really wants avoid this target of re-implementing the
constraint API it has to have its own policy (as it was in the past). I am
just not convinced that this is the best thing to do since the policy
requirements are also increasing ...

We have other places on Zephyr that we allow platforms / applications
to override some APIs and I particularly don't see it as big problem,
a bug in their implementation is not that different from another bug
affecting their code.

That's said, my main concern and reason to -1 here is just the constraint
check being done inside the power implementation, that is bad IMHO because
the rest of the subsystem has already done things assuming that that state
would be used. Also, it removes the possibility of the policy choose another
low power state.

Reply:
I agree it's not the best place, but, it's really an edge case that covers the clock readiness constraint triggered by `Power_init()`. So, when the system has booted up, this branch will in practice do nothing to my understanding. The right thing to do here, as I mention, is to implement a proper clock control driver, where Zephyr constraints can be used. `Power_init()` code is hardly exportable to an RTOS like Zephyr that already brings a PM subsystem. The solution that was taken here was to “let's defer work to HAL because platform support in Zephyr is not complete”. Platforms taking these “shortcuts” should not expect to be treated in a special way, specially TI which doesn't seem to be actively maintained. They can improve Zephyr support or make their HALs more portable.

Reply:
Some context for the discussion we had
https://github.com/zephyrproject-rtos/zephyr/issues/38362